### ISSUE TITLE
Mutagen sync taking several seconds

### ISSUE THREAD

**Issue by f1sherman** [docker/for-mac#4700]
- [x] I have tried with the latest version of my channel (Stable or Edge)
  - [x] I have uploaded Diagnostics
  - Diagnostics ID: 4C249E25-E1CB-440D-AD4E-B597FD1BBFCD/20200623142403

### Expected behavior

Filesystem changes on the host are reflected within the container within 1 second when Mutagen caching is enabled

### Actual behavior

Filesystem changes on the host are reflected within the container in ~5 seconds when Mutagen caching is enabled

### Information

Hello! I've been test driving the Mutagen sync functionality in Docker for Mac and it's been a huge improvement in I/O performance and stability. In particular, I've noticed near-native read I/O and I'm no longer seeing nested volume mounts drop as described in https://github.com/docker/for-mac/issues/3479.

One issue I'm running into is that filesystem changes on the host seem to take ~5 seconds to be reflected within the container. Based on reading through https://github.com/mutagen-io/mutagen/issues/81 I suspect that this is primarily related to some rather large dependency directories within the parent directory being sync'd. I actually don't want to sync the dependency directories and have created nested volume mounts for them as described in https://docs.docker.com/docker-for-mac/mutagen-caching/#avoiding-synchronizing-a-subdirectory. However, while filesystem changes within these dependency directories are not sync'd, they appear to still be scanned. One way to fix this might be to respect a `~/.mutagen.yml` file for ignoring specific directories. I know there are [some snapshot builds that support `~/.mutagen.yml`](https://gist.github.com/djs55/0e8957e1e36431a43cbfd3caaffa9165), but I haven't installed them since they aren't signed. If this is planned to be supported in the next build of Edge let me know and I'd be happy to wait and test that out when it's available.

- macOS Version:10.15.5
- Docker for Mac version 2.3.1.0 (45429)

### Steps to reproduce the behavior

  1. (On the host) `touch foo`
  2. (On the container) # Wait 5 seconds for `foo` to appear

**Comment by f1sherman** [comment]
It looks like I'm unable to add the `Mutagen` label as specified in the instructions here: https://docs.docker.com/docker-for-mac/mutagen-caching/#feedback

**Comment by stephen-turner** [comment]
mutagen.yml will be supported in the next Edge release, which is planned for tomorrow.

**Comment by stephen-turner** [comment]
@djs55 Any comment re the suspicion that "avoiding synchronizing a subdirectory" may not actually fully solve the problem of large subdirectories?

**Comment by stephen-turner** [comment]
Thanks for the feedback, @f1sherman.

**Comment by djs55** [comment]
I think there's still a problem, in one direction of the two-way sync.

After following the docs to exclude a directory from the sync (by creating a named `docker volume`) we have:
- all the files on the host inside the directory tree with synchronisation enabled in the UI -- minus the files excluded in the `mutagen.yml` -- are copied into the VM
- when the container is created, the named docker volumes are bind mounted on top. This means that writes within the container to the excluded directories will go to the named docker volume(s) and won't be synchronised back to the host.

So the host will still scan the excluded directories and copy the changes in, even though they can't be seen in the container. Writes inside the container will be ignored properly.

The next edge build will allow the global `mutagen.yml` to be configured to fully exclude directories from both directions of the two-way sync.

I wonder if we should support the existing `.dockerignore` file, used by `docker build`. The `.dockerignore` file could then be checked into the source code repo, which would be better than an out-of-tree global `mutagen.yml`. What do you think?

**Comment by f1sherman** [comment]
> I wonder if we should support the existing .dockerignore file, used by docker build. The .dockerignore file could then be checked into the source code repo, which would be better than an out-of-tree global mutagen.yml. What do you think?

That seems totally reasonable to me!

**Comment by f1sherman** [comment]
After updating to the latest Edge version I've had good results using `~/.mutagen.yml` to block unwanted files from syncing. It seems to have substantially improved sync times. One snag I ran into was that my initial yaml format was invalid and it took me a while to figure that out, but once I got that ironed out it started working great.

FWIW I think using `.dockerignore` would be an even better solution since it would consolidate the configuration and allow easier configuration per-project.

I'm going to go ahead and close this issue. Thank you!
